Animation and Damping

gchoqueux opened the pull request:
### New Class for animation features
* `AnimationPlayer` to control animation
* `Animation` default animation
* `AnimationExpression` to create animation with function
### Add Animation features in globe control
* double click event : it triggers a animation to zoom to mouse position
* damping moves
### Add Animation features in Api's functions
* `setCameraOrientation()`
* `setTilt()`
* `setHeading()`
* `resetTilt()`
* `resetHeading()`
* `setCenter()`
* `setCenterAdvanced()` (2 animations for moment)
* `setRange()`
* `setZoomLevel()`
* `setZoomScale()`
* `setAnimationEnabled()`
* `isAnimationEnabled()`

autra commented on index.html:
```diff
@@ -32,47 +32,52 @@
 
-            imageryPromise.then((layersColors) => menuGlobe.addImageryLayersGUI(layersColors));
-            elevationPromise.then((layersElevations) => {menuGlobe.addElevationLayersGUI(layersElevations);});
+            imageryPromise.then((layersColors) => { menuGlobe.addImageryLayersGUI(layersColors); });
```
Just chiming in because I did the linter things : is this required by the linter? (It shouldn't !)

gchoqueux replied:
yes

autra commented on index.html:
```diff
@@ -32,47 +32,52 @@
-        	});
-        </script>
+            itowns.viewer.addEventListener('globe-loaded', () => {
+                /* eslint-disable no-alert, no-console */
```
Same question: do you still have warnings/errors about `console.log`? airbnb style should allow them if I did things correctly.

And now I'm done and you guys can talk about important stuff ;-)

gchoqueux replied:
Yes, there are errors with `console.log`, this is one of the first rules we added. I prefer like this

tbroyer replied:
You can use
```js
// eslint-disable-next-line no-console
console.info('Globe Loaded');
```
instead, or
```js
console.info('Globe Loaded'); // eslint-disable-line no-console
```
(I don't think you need to disable `no-alert`)

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -44,51 +46,152 @@ var CONTROL_KEYS = {
 };
 
+// TODO: can be optimize for some uses
+var presiceSlerp = function (qb, t) {
```
Why not use Quaternion.slerp from three.js?

gchoqueux replied:
Three.js `Quaternion.slerp` considers an angle less than 0.001 to zero ([see](https://github.com/mrdoob/three.js/blob/master/src/math/Quaternion.js#L443))

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -1032,14 +1194,54 @@ GlobeControls.prototype.constructor = GlobeControls;
+GlobeControls.prototype.setOrbitalPosition = function (range, heading, tilt, isAnimated) {
+    const deltaPhi = ((tilt || this.getTilt()) * Math.PI / 180 - this.getTiltRad());
+    const deltaTheta = ((heading || this.getHeading()) * Math.PI / 180 - this.getHeadingRad());
+    const deltaRange = (range || this.getRange()) - this.getRange();
```
`range || 0` ? (the same applies to the 2 previous lines)

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+ * AnimationPlayer
+ * It can play, pause or stop animation
+ */
+function AnimationPlayer(dom) {
```
This is a very good candidate to use es6 classes :-)

autra replied:
btw AnimationPlayer is a singleton right? In this case, maybe we can just have a `getInstance` method that would construct one or return the existing one. 

Then, we can put all the private state in a closure. This way you can avoid all the `bind(this)` below by creating these functions once and for all in the closure. Then you can return a plain object with all the public methods as fields (no need to populate the prototype in this case).

gchoqueux replied:
`AnimationPlayer` isn't a singleton. 

> Then, we can put all the private state in a closure. This way you can avoid all the bind(this) below by creating these functions once and for all in the closure. Then you can return a plain object with all the public methods as fields (no need to populate the prototype in this case).

a little example?

autra replied:
> AnimationPlayer isn't a singleton.

Aren't you instanciating it only from GlobeControls's constructor?

> a little example?

Sure! if you never instanciate more than one AnimationPlayer, it could look like this

```js
// all the const here
const FRAMERATE = 60; 
// ....

// the state variables here. The one that are currently attached to `this` 
// in the constructor except this.dom
let id = null;
// ... 

// then goes all the functions, 
function resetTimer() {
    if (id) {
       clearInterval(id);
        id = undefined;
    }
    keyframe = 0;
}

function isStopped() {
    setPlayerState(PLAYER_STATE.STOP);
    finishAnimation();
    this.dom.dispatchEvent(stopEvent); // you can still use this.dom
    return promise; // I'm not sure you want a promise here
}
// ...

// then you export what you need to be visible
export default  {
    dom: null, // or document, as a default value 
    resetTimer,
    isStopped,
    //.... 
}
```
The client code just need to do `AnimationPlayer.dom = domElement;`. Here is [a working example of something similar](https://jsfiddle.net/8g68esnp/2/) (the `var myLib = (function() {...})();` emulates export and import mechanism of es6.

Imho this has several general advantages:
- what is private is *really* private: you can't access from the outside because webpack wraps all this in closures (actually you don't need to make a closure explicitly here). Btw closure is the real and only way to make things private in js. 
- it's less verbose

And additionally here, you get rid of `this`, so you don't need to `.bind` your functions (and it's faster too, because IIRC `bind` creates a new function each time it is called). The less you worry about this, the better :-)

If you need several instance of `AnimationPlayer` (but your code does not, in its current form), it's still possible, but you would need to wrap all the instance dependant fields in a closure, something like:

```js
// all the const
const ....

getInstance(dom) {
    // instanciate all instance dependant fields
    let id = ...;
    // ...
   
    // declare all the functions that depends on these field
    // they capture local variables in their closure
    function isStopped() { /* .... */ }
    return {
         isStopped,
         play,
         // ....
    };
}
export default {
    getInstance,
    // and maybe some "static" functions
}
```
I hope I'm clearer :-)

gchoqueux replied:
OK thanks for your explanations

> Aren't you instanciating it only from GlobeControls's constructor?

Yes for the moment, but there will be other uses (like earth's rotation, object3D animation...)

gchoqueux replied:
I have pushed commit : ES6 Class for Animation

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+    this.endEvent = new CustomEvent('endAnimation');
+    this.animation = null;
+    this.resolve = null;
+    this.promise = new Promise((r) => { this.resolve = r; });
```
btw, you can also write

```js
this.promise = new Promise(r => this.resolve = r);
```

Here, the `=` makes it not that nice, but when it's just a method call, it can be clearer. Just so you know :-)

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -634,33 +630,37 @@ ApiGlobe.prototype.setCenter = function (coordinates) {
-    this.setTilt(pPosition.tilt);
+ApiGlobe.prototype.setCenterAdvanced = function (pPosition, isAnimated) {
+    return this.setCenter(pPosition, isAnimated).then(() => {
+        var p = this.scene.currentControls().setOrbitalPosition(undefined, pPosition.heading, pPosition.tilt, isAnimated);
```
`const p`? or just return directly...

autra commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -44,51 +46,152 @@ var CONTROL_KEYS = {
+var animationDampingMove = new AnimatedExpression(120, ctrl, dampingMoveAnimatedExpression, 'Damping Move');
+var animationZoomCenter = new AnimatedExpression(45, ctrl, zoomCenterAnimatedExpression, 'Zoom Center');
+var animationOrbit = new AnimatedExpression(30, orbit, animationOrbitExpression, 'set Orbit');
+var OrbitDamping = new Animation(60, {}, 'orbit damping');
```
`orbitDamping` ?

autra replied:
and use const?

gchoqueux replied:
> `orbitDamping`

I change to `dampingOrbitalMvt` : damping orbital movement

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+ * @param      {Object}  root      The root object to animate
+ * @param      {String}  name      The animation's name
+ */
+function Animation(duration, root, name) {
```
Why do you need this class? Why not using only `AnimatedExpression`? (and renaming to `Animation`)

Disclaimer, I'm not fan of these inheritance techniques in javascript. They don't actually enforce anything and the language is not made for this. They really look like they are imported from Java or C++ too much to feel ok ;-)

In js, we usually use just plain object constructed with `{ .. }` and document what we expect for each functions.

autra replied:
> They don't actually enforce anything

to elaborate on that: 
- they don't enforce signature (`Object.create` does not "freeze" the object, and does not force a method to be defined. Methods can be added, removed, redefined)
- you cannot (easily) enforce type check when calling a method (so your method don't actually care to get a simple object or an object defined with new and a prototype) 

So for these reasons, you gain nothing more than if you just passed an object looking like

```js
{ 30, elem, 'name', animationFn}
```

The code would behave exactly the same, with exactly the same guarantees, but would be a lot more readable and easier to work with.

Here, Animation and AnimatedExpression are actually **data** and not **services**. They don't behave on their own, they don't need a prototype, they are just carrying some info + a callback. So I would
-  remove `Animation` and `AnimatedExpression` declarations entirely.
- replace `this.animation.animate(this.keyframe)` by `this.animation.expression(this.animation.root, keyFrame / this.animation.duration)` (it's a one-liner anyway) [on line 153](https://github.com/iTowns/itowns2/pull/218/files#diff-3629b694323d3d642bb4c6ecbebbaadbR153)
- all the AnimatedExpression declaration on GlobeControl would just be some plain objects that would be stored on `this.animation` of `AnimationPlayer` 
- Document what animation should look like in `AnimationPlayer.play`

I know this is a regular pattern in Itowns, but it's not idiomatic in js (at least a *lot* less than in Java) and we should abandon it. It is useful in other languages, but in js it's just a pain. 

Of course if I missed a reason why it's done this way, please do tell me :-)

gchoqueux replied:
> Why do you need this class?

There will be `keyFrameAnimation` inherit from `Animation` 

> `{ 30, elem, 'name', animationFn}`

Ok

> remove `Animation` and `AnimatedExpression` declarations entirely and replace `this.animation.animate(this.keyframe)` by `this.animation.expression(this.animation.root, keyFrame / this.animation.duration)`

I would prefer not to

> all the AnimatedExpression declaration on GlobeControl would just be some plain objects that would be stored on this.animation of AnimationPlayer

I don't understand

> Document what animation should look like in `AnimationPlayer.play`

Ok

autra replied:
>  >    all the AnimatedExpression declaration on GlobeControl would just be some plain objects that would be stored on this.animation of AnimationPlayer
>
> I don't understand

Just saying that in my opinion you don't need classes and inheritance to pass a 4-flelds object to you `play` method (AnimatedExpression.animate does not really add values). 

Btw inheritance does not exist in javascript per-se. Yes, you can inherit prototype, but it does not make sense to do so with empty function. 

doing
```js
Animation.prototype.animate = function () {};
AnimatedExpression.prototype = Object.create(Animation.prototype);
AnimatedExpression.prototype.animate = function (keyFrame) {    /* actually do something */ };
```

is the same as doing
```js
AnimatedExpression.prototype.animate = function() {};
AnimatedExpression.prototype.animate = function(keyFrame) { /* do the real thing */ };
```
It's not very useful :-) You don't need `Animation`.

gchoqueux replied:
but i use Animation [here](https://github.com/iTowns/itowns2/blob/05fa85953e12d4f9691e48bcde5e07d850f659c8/src/Renderer/ThreeExtented/GlobeControls.js#L182)

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+
+Animation.prototype = Object.create(Animation.prototype);
+
+Animation.prototype.animate = function () {
```
Following my previous comment: especially because this construction requires making empty function just *anywhere*.

gchoqueux replied:
I have removed `Animation.prototype.animate`

autra commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -958,13 +1108,22 @@ function GlobeControls(camera, domElement, engine) {
     };
 
+    // Callback launched when player is stopped
+    this.stopAnimation = function () {
```
If it is to be called each time an animation is finished, please consider having this callback as a parameter when constructing the `AnimationPlayer`. This way you won't need to add
```js
player.play().then(() => this.stopAnimation())
```
Or something similar each time. Then this callback would be called each time you call `this.resolve` in AnimationPlayer.

Btw, promises are not made for recurring events. They are made for one-time resolvable event, for which they are great. But as soon as your need can be expressed as

> I want to do A every time B happens

They are poor candidates. Callbacks are still the right solution for this.

gchoqueux replied:
ok

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+    setPlayerState.bind(this)(PLAYER_STATE.STOP);
+    finishAnimation.bind(this)();
+    this.dom.dispatchEvent(this.stopEvent);
+    return this.promise;
```
So I have a problem with `this.promise`:

- you init it in the constructor and capture the resolve function in `this.resolve`
- you return the promise only in `stop`
- you change `this.resolve` in `play` but not `this.promise`
- you resolve it in `finishAnimation`

As I expect `play` to be called before `stop` or `finishAnimation`, I don't see how `this.promise` could ever be resolved...

gchoqueux replied:
I rectify

autra reviewed (commented):
Just passing by for my own culture, leaving some remarks. It's in no way exhaustive :-)

gchoqueux commented:
@autra I push new commit
When using, does the animation work well?

autra commented:
>  When using, does the animation work well?

Yes, except that I still have the bug when everything goes upside-down from time-to-time (but I had that before your patch)

Damping is the continuation of the movement when you release the mouse right? 

If so, it works only one way with the ctrl key (when increasing the angle diff to the normal), but not the other way (when going back near the normal), is it what you wanted? I can't trigger any damping with any other actions (but maybe that's intended). It's a bit laggy at the beginning of the animation though.. I'm testing on firefox

autra commented:
> If so, it works only one way with the ctrl key (when increasing the angle diff to the normal), but not the other way (when going back near the normal), is it what you wanted? I can't trigger any damping with any other actions (but maybe that's intended). It's a bit laggy at the beginning of the animation though.. I'm testing on firefox

Correction, things seems to be working as expected. I don't why I couldn't trigger the damping in my first tests.

gchoqueux commented:
> Damping is the continuation of the movement when you release the mouse right?

Yes

gchoqueux commented:
@autra I have fix all your comments
and add ES6 Class for Animation

gchoqueux commented:
Rebased and Ready

gchoqueux committed:
refactor: getRTCMatrixFromCenter
c3DEngine fix error picking
Clean up code

gchoqueux committed:
feat: Add animation feature

gchoqueux committed:
feat: Add animation option in api functions

gchoqueux commented:
Updated

gchoqueux committed:
Upgrade Three.js v82 to v83

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+    ];
+
+// if debugAnimation is true console.log are enabled to sniff animation'state
+var _DEBUG = null;
```
nits: let

autra replied:
And there are other occurrences of var below. It's not critical though, they behave nearly the same. But let has a Temporal Dead Zone, which is useful to fail fast.

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+const FRAMERATE = 60;
+const INTERVAL_TIME = 1000 / FRAMERATE;
+// if is true console.log are enabled to sniff animation'state
+const debugAnimation = false;
```
Maybe we can have a global that would put us into debug mode. It would be `false` by default and could be made true on dev mode. Then we could have a `debug.js` class that would contain our _DEBUG (or we could overwrite console.info). It would avoid to have a different debug var for each class.

I'll try to make a PR with this maybe this week or the next

gchoqueux replied:
Ok global debug with options (shader, animation, tile, command...)

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+     * { this function is executed with each frame }
+     */
+    frame() {
+        if (this.keyframe === this.animation.duration) {
```
Maybe 
```js
if (this.keyframe < this.animation.duration) {
    // code in current else
} else {
    // code in current if
}
```
It would allow to skip frames without risks.

autra commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,203 @@
+        this.keyframe = 0;
+        this.frameEvent = new CustomEvent('frameAnimation');
+        this.stopEvent = new CustomEvent('stopAnimation');
+        this.endEvent = new CustomEvent('endAnimation');
```
So when you fire an event like this, it gets a bunch of other attributes (`target`, `currentTarget`, `cancelable`...) from the browser and then is given to event listeners. For this reason, I'm not very comfortable with having static ones like that. I think it's safer to instanciate a new CustomEvent each time you dispatch it. It would also reduce the number of fields we have here (at the very least, these should be module constants).

autra commented on package.json:
```diff
@@ -29,7 +29,7 @@
     "simd": "^2.0.0",
     "string_format": "^0.0.6",
-    "three": "^0.82.0",
+    "three": "^0.83.0",
```
This could be merged immediately if it has its own PR :-)

autra commented on index.html:
```diff
@@ -32,47 +32,51 @@
+                // eslint-disable-next-line no-console
+                console.info('Globe Loaded');
+            });
+
```
All this trivial change (space, formatting etc...) would be merged very quickly if they had their own PR. It would also make this one easier to review imho.

gchoqueux replied:
yes

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -613,17 +630,38 @@ ApiGlobe.prototype.computeDistance = function computeDistance(p1, p2) {
     return this.scene.getEllipsoid().computeDistance(new GeoCoordinate().copy(p1), new GeoCoordinate().copy(p2));
 };
 
+ApiGlobe.prototype.setSceneLoaded = function computeDistance(callback) {
```
Why is the function name `computeDistance`?

gchoqueux replied:
oops

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -613,17 +630,38 @@ ApiGlobe.prototype.computeDistance = function computeDistance(p1, p2) {
 
+ApiGlobe.prototype.setSceneLoaded = function computeDistance(callback) {
+    loaded = false;
+    this.callback = () => {
```
If you call 2 times `setSceneLoaded` before `globe-loaded` fires, only one of the 2 callbacks would get called. Actually, if you want to execute some piece of code once after the next `globe-loaded` event, promises are a good way to do so. More precisely, you can use what is commonly called **deferred**:

They are built this way (they are just wrappers around promise and their reject and resolve functions).
```js
// in some utils module maybe?
function defer() {
    let defer = {};
    defer.promise = new Promise((resolve, reject) => {
        defer.resolve = resolve;
        defer.reject = reject;
    });
    return defer;
}
```
Then instead of `this.callback` you will instanciate a defer, fire and replace it each time globe-loaded fires:
```js
this.globeLoadedDeferred = defer(); // precise names are better
this.addEventListener('globe-loaded', () => {
    this.globeLoadedDeferred.resolve();
    this.globeLoadedDeferred = defer();
});
```
And each time you currently call `setSceneLoaded`:
```js
this.globeLoadedDeferred.promise.then(<your callback>);
```
This way, each registered  callbacks since last `globe-loaded` would get called the next time the event fires. No need to add and remove listeners.

Also, by returning promises, you can chain them (see below).

gchoqueux replied:
Nice

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -613,17 +630,38 @@ ApiGlobe.prototype.computeDistance = function computeDistance(p1, p2) {
-    this.viewerDiv.dispatchEvent(eventCenter);
+    const position3D = this.scene.getEllipsoid().cartographicToCartesian(new GeoCoordinate(coordinates.longitude, coordinates.latitude, 0, UNIT.DEGREE));
+    position3D.range = coordinates.range;
+    return new Promise((r) => {
```
With the above change,  this will become
```js
    return this.scene.currentControls().setCenter(position3D, isAnimated).then(() => {
        this.scene.notifyChange(1);
        return this.globeLoadedDeferred.promise.then(() => {
            this.scene.currentControls().updateCameraTransformation();
            this.viewerDiv.dispatchEvent(eventCenter);
        });
    });
```
And calls to `setCenter().then` would also get executed after this then.

autra replied:
Less boilerplate, isn't it?

gchoqueux replied:
thanks

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -634,33 +672,39 @@ ApiGlobe.prototype.setCenter = function setCenter(coordinates) {
-    this.scene.currentControls().setRange(pRange);
-    this.addEventListener('globe-loaded', updateTargetCamera(this));
+
+    return new Promise((r) => {
```
Same modifications as in `setCenter`

autra reviewed (commented):
I have not finished my review, but as I won't be available the next 3 days, I'm submitting this part right now. I'll get back to it asap.

gchoqueux commented:
updated

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,204 @@
+ * @param      {Number}  duration  The animation's duration
+ * @param      {Object}  root      The root object to animate
+ * @param      {String}  name      The animation's name
+ */
```
10 lines of periphrases do not make a good documentation.

If you add documentation, make it useful:
* explain the purpose of Animation: move objects? ik based animation? lerp between values? generic?
* specify units (duration is in seconds? days? µs ?)
* why|when should I use (not) `Animation`?

peppsac replied:
btw, this comment applies to all the new comments added in this PR.

Documentation is hard but code review helps improving it.

peppsac replied:
Last comment about comments: where does this strange layout (`{ Animation `} instead of `Animation`) comes from?

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,204 @@
+        this.callBack = callBack;
+        this.animation = animation;
+        if (this.animation.animate) this.callBackFrame = () => this.animation.animate(this.keyframe);
+        setPlayerState.bind(this)(PLAYER_STATE.PLAY);
```
I'm not sure why there are so much `bind(this)` in this code ?

gchoqueux replied:
I remove the `bind(this)`

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,204 @@
+};
+
+const setPlayerState = function setPlayerState(state) {
+    this.state = state;
```
```js
const setPlayerState = function setPlayerState(player, state) {
   player.state = state;
```
is less clumsy and doesn't require a `bind(this)`

gchoqueux replied:
yes

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,204 @@
+     * @return     {Promise}  Promise is resolved when animation is stopped or finished
+     */
+    play(animation, callBack) {
+        this.callBack = callBack;
```
Why do you mix `callbacks` and `promises`?
IMHO you should drop `this.callBack` unless you have a strong argument

gchoqueux replied:
@autra in this [comment](https://github.com/iTowns/itowns2/pull/218#discussion_r92003633) recommended to use callback

But Effectively the promise isn't used in play function

autra replied:
What @peppsac means is that `this.callback` and `this.promise` are redundant to each other (they are both called/fired at the end of the animation). I suggested to use callback for each frame. I guess you need only `this.promise` and `this.callbackFrame` right?

gchoqueux replied:
> I suggested to use callback for each frame

This has been the case since the beginning, [it was for callback in animation's ending](https://github.com/iTowns/itowns2/pull/218#discussion_r92003633). I see this

autra replied:
I don't know why I said that earlier and I can't remember how the code was at this time. We might have misunderstood each other :-)

So anyway: one promise for the end of the animation and a callback for each frame. Sounds good to you guys?

gchoqueux replied:
> one promise for the end of the animation and a callback for each frame. Sounds good to you guys?

Yes, I revert to this code

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,204 @@
+        this.keyframe = 0;
+        this.animation = null;
+        this.resolve = null;
+        this.promise = new Promise(r => this.resolve = r);
```
Not needed

gchoqueux replied:
How would you initialize it?

gchoqueux commented:
updated

autra commented:
Ok I think I see where the misunderstanding lies.

I think AnimationPlayer.promise is not at the right place. It should be in Animation (and won't ever be null, you init it when you create the animation, resolve it when it's finished). It's a state information about an animation, not the player. Same goes for callbackFrame, it's an information about Animation, not AnimationPlayer. And same goes for state. Anyway, I think everything would be so much simpler if your animations could play themselves (so actually if AnimationPlayer and Animation would be merged). 

```js
let animation = new Animation(..., callbackFrame);
animation.play().then(...).catch(...);
...
// later
animation.cancel(); // catch is executed 
// or
animation.pause();
animation.resume();
```

gchoqueux commented:
> so actually if AnimationPlayer and Animation would be merged

Later, I would like to evolve `AnimationPlayer` to play animations simultaneously.

```js
AnimationPlayer.play(animation_A, animation_B);
```
To synchronize the playing, it would take a manager like the AnimationPlayer

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+        return INTERVAL_TIME;
+    }
+
+    isPlayed() {
```
isPlaying

gchoqueux replied:
ok

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+     * @return {Number}  The interval time is the time between consecutive frames
+     */
+    static get intervalTime() {
+        return INTERVAL_TIME;
```
* `frameDuration` would be a better name
* you don't know the framerate so assuming it'll be 60 FPS might be wrong...
* last but not least, this function is unused so you could remove it

gchoqueux replied:
I remove it

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+const FRAMERATE = 60;
+const INTERVAL_TIME = 1000 / FRAMERATE;
```
FRAME_DURATION ?

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+     * @param      {Animation} The animation to play
+     * @return     {Promise}  Promise is resolved when animation is stopped or finished
+     */
+    play(animation) {
```
AnimationPlayer only support 1 animation at a time, so you must check the value of `this.state`

gchoqueux replied:
If another animation is playing, it's stopped and the new animation is played
(comment add in code)

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+// player statut
+const PLAYER_STATE = {
+    // player is stopped
+    STOP: 0,
```
What's the difference between `STOP` and `PAUSE`? (there's no `resume` function)

gchoqueux replied:
I add resume()

peppsac replied:
I'd prefer to remove `PAUSE` state until there's a real need for it.

gchoqueux replied:
ok

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -651,121 +765,155 @@ function GlobeControls(camera, domElement, engine) {
+                player.play(dampingOrbitalMvt).then(() => this.stopAnimation());
+            } else if (state === CONTROL_STATE.MOVE_GLOBE && lastRotation.length === 2 && !(lastRotation[1].equals(lastRotation[0]))) {
+                ctrl.qDelta.setFromUnitVectors(lastRotation[1], lastRotation[0]);
+                player.play(animationDampingMove).then(() => this.stopAnimation());
```
Why do you stop the animation after it ended?

gchoqueux replied:
it's bad name... 
I rename `reinitiateControls`, this function reinitiates somes controls's params.

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+     * @param      {Number}  waitingTime  The waiting time before start animation (time in frame)
+     */
+    playLater(animation, waitingFrame) {
+        this.resolveWait = null;
```
Check `this.state`

peppsac replied:
(this comment also applies to `stop` and `pause` methods)

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+        if (this.keyframe < this.animation.duration) {
+            this.callBackFrame();
+            this.keyframe++;
+            this.dom.dispatchEvent(frameEvent);
```
`event` and `callback` provide the same feature.

Why don't you drop one (callback)?

gchoqueux replied:
`callBackFrame` is function to animate object
`frameEvent` is used to resfresh display.

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,231 @@
+ */
+class Animation {
+    constructor(params) {
+        this.duration = params.duration || 0;
```
Maybe pick a more sensible default value, 60 (1 sec) ? Or throw an error if `params.duration` is unset.

peppsac reviewed (changes requested):
While the animation works, there are still a few things to fix before merging.

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -958,13 +1104,22 @@ function GlobeControls(camera, domElement, engine) {
     };
 
+    // Callback launched when player is stopped
+    this.reinitiateControls = function reinitiateControls() {
```
`resetControls` ?

peppsac commented on src/Scene/AnimationPlayer.js:
```diff
@@ -0,0 +1,214 @@
+     */
+    frame() {
+        if (this.keyframe < this.animation.duration) {
+            this.callBackFrame();
```
I'd suggest removing `this.callBackFrame` and simply do:
```js
if (animation.animate) {
    animation.animate(this.keyframe;
}
```

gchoqueux committed:
feat: Add damping in globe's control, in state MOVE_GLOBE and ORBIT
Add dblClick feature : move camera to click point
Best behavior in mouse's move when cursor is outside the globe
